proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,649 @@
+
+    f :: forall a.   Tagged a ()
+    g :: forall a -> Tagged a ()
+
+    data T = T
+
+    a = f @T
+    b = g  T
+
+  In ``f @T``, we refer to the type constructor, but in ``g T`` we refer to the
+  data constructor.
+
+  The implementation may offer warning flags to help the user identify such
+  ambiguous occurrences.
+
+* When punned names come from external libraries, there are several workarounds
+  to resolve the ambiguity:
+
+  1. Using ``-XExplicitNamespaces``::
+
+      import Data.Proxy
+      import qualified Data.Proxy (type Proxy) as T
+
+      x = f   Proxy  -- refers to the data constructor
+      y = f T.Proxy  -- refers to the type constructor
+
+  2. Using a type synonym::
+
+      import Data.Proxy
+
+      type TProxy = Proxy
+
+      x = f  Proxy  -- refers to the data constructor
+      y = f TProxy  -- refers to the type constructor
+
+  We may offer a more convenient way to disambiguate namespaces in the future,
+  based on the experience of using the proposed feature.
+
+* Identifiers bound to terms are
+  not promoted. Consider this well-typed program::
+
+    f :: forall a.   Tagged a ()
+    g :: forall a -> Tagged a ()
+
+    a = f @(Just True)  -- ok
+    b = g  (Just True)  -- ok
+
+  If we factor out ``Just True`` into a type synonym, it continues to work::
+
+    type X :: Maybe Bool
+    type X = Just True
+
+    a = f @X  -- still OK
+    b = g  X  -- still OK
+
+  However, if we bind it to a term-level variable, the example becomes
+  ill-typed::
+
+    x :: Maybe Bool
+    x = Just True
+
+    a = f @x  -- not currently valid
+    b = g  x  -- not valid under the proposal
+
+  This is because we retain the distinction between terms and types. This
+  proposal is a step towards dependent types, but it does not go all the way.
+  Accepting the program above is left as future work.
+
+Costs and Drawbacks
+-------------------
+
+This is one more feature to implement and maintain.
+
+Alternatives
+------------
+
+1.  Keep types and terms forever different by not supporting visible ``forall``
+    in terms.
+
+2.  Include the proposed functionality in ``ExplicitForAll`` instead of
+    introducing a new extension.
+
+3.  The extension name could use a different name, such as ``-XVDQ`` or
+    ``-XVisibleForAll``.
+
+4.  We could guard type-level uses of visible ``forall`` behind the
+    ``VisibleForAll`` extension flag. This would break existing code.
+
+5.  To avoid the name resolution issues, we could introduce a syntactic marker to
+    indicate visible type arguments. That is, for some ``f :: forall x
+    -> ...``, one would pass ``x`` as ``f ^x`` instead of ``f x``, where ``^``
+    is new syntax (and a different choice of symbols is possible). There are
+    several issues with this alternative:
+
+    * it creates more syntactic noise
+    * it is inconsistent with what we have in types where we do not need a marker
+    * it does not move us towards pi-types
```

Comment:
This claim needs to be substantiated: how would requiring such a marker be less of a move in the direction of pi types than not requiring one?

Reply:
It assumes that we don't want a marker for pi-types, so the work done here on parsing and name resolution will carry over when we start implementing dependent types. So when it comes to dependent relevant quantification, we will only need to worry about the 'relevant' part, not about the syntax minutiae.

Reply:
I'm afraid I don't know what you mean.

Reply:
I expanded the sentence into:

> it does not move us towards pi-types, assuming we don't want function application of dependent functions to also use a marker

I will try to elaborate. My understanding is that pi-types, as described in "Dependent Types in Haskell: Theory and Practice" entail a complete merger between the type level and the term level, so that the current typing relation `e :: t` and kinding relation `t :: k` are, in fact, the same relation.

Therefore, if `Int :: Type` is valid at the type level, it is also valid at the term level. And we could write a dependent, term-level function `f` that would be used as `f (Maybe Int -> Bool)`. However, we don't want to write this function application as `f ^(Maybe Int -> Bool)`, or else it means that function application and dependent function application have different syntax, and that would be an ugly wart on the look & feel of Dependent Haskell, which would make it very unhaskell-y.

So there are three steps to make pi-types work:

1. Merge the syntax of types and terms
2. Merge the scoping rules of types and terms
3. Merge kind checking and type checking

This proposal addresses (1) and (2), so that we only need to worry about (3) when we get to pi-types.